File: scripts/git-report-examples.mjs
/**
 * Commit examples for git-report.sh --report examples.
 * Reads git history from cwd (repo root). Used by MCP / npm run report:git.
 */
import { spawnSync } from 'node:child_process';
import { parseArgs } from 'node:util';

function runGit(args, cwd) {
  const r = spawnSync('git', args, {
    cwd,
    encoding: 'utf8',
    maxBuffer: 32 * 1024 * 1024
  });
  if (r.status !== 0) {
    process.stderr.write(r.stderr || r.stdout || `git ${args.join(' ')} failed\n`);
    process.exit(r.status ?? 1);
  }
  return r.stdout;
}

function conventionalType(subject) {
  const m = subject.match(/^([a-z]+)(?:\([^)]*\))?:/);
  return m ? m[1] : null;
}

function conventionalScopes(subject) {
  const m = subject.match(/^[a-z]+\(([^)]*)\):/);
  if (!m) {
    return [];
  }
  return m[1]
    .split(',')
    .map(s => s.trim())
    .filter(Boolean);
}

function scopeMatches(subject, want) {
  const wantL = want.toLowerCase();
  return conventionalScopes(subject).some(s => s.toLowerCase() === wantL);
}

function typeMatches(subject, want) {
  if (!want) {
    return true;
  }
  const t = conventionalType(subject);
  return t ? t === want.toLowerCase() : false;
}

/** Case-insensitive glob with * wildcard (aligned with Python fnmatch for * segments). */
function subjectGlobMatch(subject, pattern) {
  const s = subject.toLowerCase();
  const p = pattern.toLowerCase();
  const segs = p.split('*').map(seg => seg.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'));
  return new RegExp(`^${segs.join('.*')}$`).test(s);
}

function suggestedPatterns(subject) {
  const s = subject.toLowerCase();
  const out = [];
  if (s.startsWith('build(deps)') && s.includes('bump')) {
    out.push(
      'Grouped dependency bump: build(deps) subject, bullet body listing ' +
        'package from X to Y, lockfile + occasional snapshot updates.'
    );
  }
  if (s.includes('patternfly') || s.includes('@patternfly')) {
    out.push(
      'PatternFly-related change: often paired with @patternfly/* version lines ' +
        'in the commit body for bumps.'
    );
  }
  if (/^refactor\(deps\):/.test(subject) && s.includes('pf')) {
    out.push(
      'Major PF migration: refactor(deps) with tracker id, long bullet list of ' +
        'component-level adjustments.'
    );
  }
  if (/^fix(\([^)]*\))?:/.test(subject)) {
    out.push('Fix commit: fix(scope): or fix: with optional tracker/PR suffix.');
  }
  return out;
}

const { values } = parseArgs({
  args: process.argv.slice(2),
  options: {
    'as-of-commit': { type: 'string' },
    'as-of-date': { type: 'string' },
    since: { type: 'string' },
    until: { type: 'string' },
    limit: { type: 'string', default: '15' },
    type: { type: 'string' },
    scope: { type: 'string' },
    'subject-glob': { type: 'string' },
    path: { type: 'string', multiple: true },
    format: { type: 'string', default: 'json' },
    'body-lines': { type: 'string', default: '12' }
  },
  strict: true,
  allowPositionals: false
});

const asOfCommit = values['as-of-commit'];
const asOfDate = values['as-of-date'];
if (!asOfCommit || !asOfDate) {
  process.stderr.write('Missing --as-of-commit or --as-of-date\n');
  process.exit(1);
}

const cwd = process.cwd();
const rev = asOfCommit;
const limit = Math.min(100, Math.max(1, parseInt(values.limit, 10) || 15));
const bodyLines = Math.max(0, parseInt(values['body-lines'], 10) || 12);
const ctype = values.type ?? '';
const scope = values.scope ?? '';
const subjectGlob = values['subject-glob'] ?? '';
const since = values.since ?? '';
const until = values.until ?? '';
const paths = values.path ?? [];
const format = values.format === 'md' ? 'md' : 'json';

const logCmd = ['log', rev, '--no-merges', '--format=%H\t%s\t%aI'];
if (since) {
  logCmd.push('--since', since);
}
if (until) {
  logCmd.push('--until', until);
}
if (paths.length) {
  logCmd.push('--', ...paths);
}

const raw = runGit(logCmd, cwd).trim();
const rows = [];
for (const line of raw.split('\n')) {
  const t1 = line.indexOf('\t');
  const t2 = t1 === -1 ? -1 : line.indexOf('\t', t1 + 1);
  if (t1 === -1 || t2 === -1) {
    continue;
  }
  const h = line.slice(0, t1);
  const subj = line.slice(t1 + 1, t2);
  const adate = line.slice(t2 + 1);
  if (ctype && !typeMatches(subj, ctype)) {
    continue;
  }
  if (scope && !scopeMatches(subj, scope)) {
    continue;
  }
  if (subjectGlob && !subjectGlobMatch(subj, subjectGlob)) {
    continue;
  }
  rows.push([h, subj, adate]);
}

const totalMatches = rows.length;
const slice = rows.slice(0, limit);

const examples = [];
for (const [sha, subject, adate] of slice) {
  const body = runGit(['log', '-1', '--format=%b', sha], cwd).trim();
  const allBodyLines = body ? body.split('\n') : [];
  const bl = allBodyLines.slice(0, bodyLines);
  let bodyExcerpt = bl.join('\n');
  if (bodyExcerpt && allBodyLines.length > bl.length) {
    bodyExcerpt += '\n…';
  }

  const statOut = runGit(['show', '--no-patch', '--stat', sha], cwd);
  const statLines = statOut.trim().split('\n');
  const statSummary = statLines.length ? statLines[statLines.length - 1] : '';

  examples.push({
    sha,
    date: adate,
    subject,
    body_excerpt: bodyExcerpt,
    stat_summary: statSummary,
    suggested_patterns: suggestedPatterns(subject)
  });
}

const filters = {
  since: since || null,
  until: until || null,
  limit,
  type: ctype || null,
  scope: scope || null,
  subject_glob: subjectGlob || null,
  paths: paths.length ? paths : null
};

if (format === 'json') {
  process.stdout.write(
    JSON.stringify(
      {
        as_of_commit: asOfCommit,
        as_of_date: asOfDate,
        filters,
        match_count: totalMatches,
        returned: examples.length,
        examples
      },
      null,
      2
    ) + '\n'
  );
  process.exit(0);
}

const lines = [
  '# Commit examples',
  '',
  '| Field | Value |',
  '|-------|-------|',
  `| **as_of_commit** | \`${asOfCommit}\` |`,
  `| **as_of_date** | ${asOfDate} |`,
  `| **match_count** | ${totalMatches} |`,
  `| **returned** | ${examples.length} |`,
  '',
  '## Filters',
  '',
  '```json',
  JSON.stringify(filters, null, 2),
  '```',
  ''
];

for (const ex of examples) {
  lines.push(`### \`${ex.sha.slice(0, 7)}\` — ${ex.subject}`, '');
  lines.push(`- **date:** ${ex.date}`);
  lines.push(`- **stat:** ${ex.stat_summary}`);
  lines.push('');
  if (ex.suggested_patterns.length) {
    lines.push('**Patterns:**');
    for (const sp of ex.suggested_patterns) {
      lines.push(`- ${sp}`);
    }
    lines.push('');
  }
  if (ex.body_excerpt) {
    lines.push('```');
    lines.push(ex.body_excerpt);
    lines.push('```');
    lines.push('');
  }
}

process.stdout.write(lines.join('\n'));

File: scripts/git-report-examples.py
#!/usr/bin/env python3
"""
Commit examples for git-report.sh --report examples.
Reads git history from cwd (repo root). Used by MCP / npm run report:git.
"""
from __future__ import annotations

import argparse
import json
import re
import subprocess
import sys
from fnmatch import fnmatch
from pathlib import Path


def _run_git(args: list[str], cwd: Path) -> str:
    r = subprocess.run(
        ["git", *args],
        cwd=cwd,
        capture_output=True,
        text=True,
        check=False,
    )
    if r.returncode != 0:
        sys.stderr.write(r.stderr or r.stdout or f"git {' '.join(args)} failed\n")
        sys.exit(r.returncode or 1)
    return r.stdout


def _conventional_type(subject: str) -> str | None:
    m = re.match(r"^([a-z]+)(?:\([^)]*\))?:", subject)
    return m.group(1) if m else None


def _conventional_scopes(subject: str) -> list[str]:
    m = re.match(r"^[a-z]+\(([^)]*)\):", subject)
    if not m:
        return []
    return [s.strip() for s in m.group(1).split(",") if s.strip()]


def _scope_matches(subject: str, want: str) -> bool:
    want_l = want.lower()
    for s in _conventional_scopes(subject):
        if s.lower() == want_l:
            return True
    return False


def _type_matches(subject: str, want: str | None) -> bool:
    if not want:
        return True
    t = _conventional_type(subject)
    return t == want.lower() if t else False


def _suggested_patterns(subject: str) -> list[str]:
    s = subject.lower()
    out: list[str] = []
    if s.startswith("build(deps)") and "bump" in s:
        out.append(
            "Grouped dependency bump: build(deps) subject, bullet body listing "
            "package from X to Y, lockfile + occasional snapshot updates."
        )
    if "patternfly" in s or "@patternfly" in s:
        out.append(
            "PatternFly-related change: often paired with @patternfly/* version lines "
            "in the commit body for bumps."
        )
    if re.match(r"^refactor\(deps\):", subject) and "pf" in s:
        out.append(
            "Major PF migration: refactor(deps) with tracker id, long bullet list of "
            "component-level adjustments."
        )
    if re.match(r"^fix(\([^)]*\))?:", subject):
        out.append("Fix commit: fix(scope): or fix: with optional tracker/PR suffix.")
    return out


def main() -> None:
    p = argparse.ArgumentParser(description="Export commit examples as JSON or Markdown.")
    p.add_argument("--as-of-commit", required=True)
    p.add_argument("--as-of-date", required=True)
    p.add_argument("--since", default="")
    p.add_argument("--until", default="")
    p.add_argument("--limit", type=int, default=15)
    p.add_argument("--type", dest="ctype", default="", help="Conventional type, e.g. fix, build")
    p.add_argument("--scope", default="", help="Conventional scope (exact match in scope list)")
    p.add_argument(
        "--subject-glob",
        default="",
        help="Case-insensitive glob on subject (* wildcard), e.g. *patternfly*",
    )
    p.add_argument(
        "--path",
        action="append",
        default=[],
        dest="paths",
        help="Pathspec (repeatable); limits to commits touching these paths",
    )
    p.add_argument("--format", choices=("json", "md"), default="json")
    p.add_argument(
        "--body-lines",
        type=int,
        default=12,
        help="Max body lines per example in JSON/md output",
    )
    args = p.parse_args()

    cwd = Path.cwd()
    limit = max(1, min(args.limit, 100))
    rev = args.as_of_commit

    log_cmd = [
        "log",
        rev,
        "--no-merges",
        "--format=%H\t%s\t%aI",
    ]
    if args.since:
        log_cmd.extend(["--since", args.since])
    if args.until:
        log_cmd.extend(["--until", args.until])
    if args.paths:
        log_cmd.append("--")
        log_cmd.extend(args.paths)

    raw = _run_git(log_cmd, cwd).strip()
    rows: list[tuple[str, str, str]] = []
    for line in raw.splitlines():
        parts = line.split("\t", 2)
        if len(parts) < 3:
            continue
        h, subj, ad = parts[0], parts[1], parts[2]
        if args.ctype and not _type_matches(subj, args.ctype):
            continue
        if args.scope and not _scope_matches(subj, args.scope):
            continue
        if args.subject_glob and not fnmatch(subj.lower(), args.subject_glob.lower()):
            continue
        rows.append((h, subj, ad))

    total_matches = len(rows)
    rows = rows[:limit]

    examples: list[dict] = []
    for sha, subject, adate in rows:
        body = _run_git(["log", "-1", "--format=%b", sha], cwd).strip()
        body_lines = body.splitlines()[: max(0, args.body_lines)] if body else []
        body_excerpt = "\n".join(body_lines)
        if body_excerpt and len(body.splitlines()) > len(body_lines):
            body_excerpt += "\n…"

        stat_out = _run_git(["show", "--no-patch", "--stat", sha], cwd)
        stat_lines = stat_out.strip().splitlines()
        stat_summary = stat_lines[-1] if stat_lines else ""

        examples.append(
            {
                "sha": sha,
                "date": adate,
                "subject": subject,
                "body_excerpt": body_excerpt,
                "stat_summary": stat_summary,
                "suggested_patterns": _suggested_patterns(subject),
            }
        )

    filters = {
        "since": args.since or None,
        "until": args.until or None,
        "limit": limit,
        "type": args.ctype or None,
        "scope": args.scope or None,
        "subject_glob": args.subject_glob or None,
        "paths": args.paths if args.paths else None,
    }

    if args.format == "json":
        out = {
            "as_of_commit": args.as_of_commit,
            "as_of_date": args.as_of_date,
            "filters": filters,
            "match_count": total_matches,
            "returned": len(examples),
            "examples": examples,
        }
        print(json.dumps(out, indent=2))
        return

    # md
    lines = [
        "# Commit examples",
        "",
        f"| Field | Value |",
        f"|-------|-------|",
        f"| **as_of_commit** | `{args.as_of_commit}` |",
        f"| **as_of_date** | {args.as_of_date} |",
        f"| **match_count** | {total_matches} |",
        f"| **returned** | {len(examples)} |",
        "",
        "## Filters",
        "",
        "```json",
        json.dumps(filters, indent=2),
        "```",
        "",
    ]
    for ex in examples:
        lines.extend(
            [
                f"### `{ex['sha'][:7]}` — {ex['subject']}",
                "",
                f"- **date:** {ex['date']}",
                f"- **stat:** {ex['stat_summary']}",
                "",
            ]
        )
        if ex["suggested_patterns"]:
            lines.append("**Patterns:**")
            for sp in ex["suggested_patterns"]:
                lines.append(f"- {sp}")
            lines.append("")
        if ex["body_excerpt"]:
            lines.append("```")
            lines.append(ex["body_excerpt"])
            lines.append("```")
            lines.append("")
    sys.stdout.write("\n".join(lines))


if __name__ == "__main__":
    main()
